src/execution/physical_plan/plan_comparison_join.cpp

```diff
@@ -29,6 +32,80 @@ static bool CanPlanIndexJoin(Transaction &transaction, TableScanBindData *bind_d
 	return true;
 }
 
+void CheckForPerfectJoinOpt(LogicalComparisonJoin &op, PerfectHashJoinStats &join_state) {
+	// we only do this optimization for inner joins
+	if (op.join_type != JoinType::INNER) {
+		return;
+	}
+	// with one condition
+	if (op.conditions.size() != 1) {
+		return;
+	}
+	// with propagated statistics
+	if (op.join_stats.empty()) {
+		return;
+	}
+	// with equality condition and null values not equal
+	for (auto &&condition : op.conditions) {
+		if (condition.comparison != ExpressionType::COMPARE_EQUAL || condition.null_values_are_equal) {
+			return;
+		}
+	}
+	// with integral types
+	for (auto &&join_stat : op.join_stats) {
+		if (!TypeIsInteger(join_stat->type.InternalType()) || join_stat->type.InternalType() == PhysicalType::INT128) {
+			// perfect join not possible for non-integral types or hugeint
+			return;
+		}
+	}
+
+	// and when the build range is smaller than the threshold
+	auto stats_build = reinterpret_cast<NumericStatistics *>(op.join_stats[0].get()); // lhs stats
+	if (stats_build->min.is_null || stats_build->max.is_null) {
+		return;
+	}
+	auto build_range = stats_build->max - stats_build->min; // Join Keys Range
+
+	// Fill join_stats for invisible join
+	auto stats_probe = reinterpret_cast<NumericStatistics *>(op.join_stats[1].get()); // rhs stats
+
+	// The max size our build must have to run the perfect HJ
+	const idx_t MAX_BUILD_SIZE = 1000000;
+	join_state.probe_min = stats_probe->min;
+	join_state.probe_max = stats_probe->max;
+	join_state.build_min = stats_build->min;
+	join_state.build_max = stats_build->max;
+	join_state.estimated_cardinality = op.estimated_cardinality;
+	if (build_range.type().id() == LogicalTypeId::DECIMAL) {
+		switch (build_range.type().InternalType()) {
+		case PhysicalType::INT16:
+			join_state.build_range = build_range.value_.smallint;
+			break;
+		case PhysicalType::INT32:
+			join_state.build_range = build_range.value_.integer;
+			break;
+		case PhysicalType::INT64:
```

Comment:
Decimals can also be HUGEINT

Reply:
True, but that physical type is ignored for this join, I can add it anyway.